fili-core/src/main/java/com/yahoo/bard/webservice/data/filterbuilders/DruidInFilterBuilder.java

```diff
@@ -48,28 +47,32 @@ protected Filter buildDimensionFilter(Dimension dimension, Set<ApiFilter> filter
             throws DimensionRowNotFoundException {
         LOG.trace("Building dimension filter using dimension: {} \n\n and set of filter: {}", dimension, filters);
 
-        List<Filter> inFilters = new ArrayList<>();
+        List<Filter> inFilters = new ArrayList<>(); // A list with at most 2 in-filters(positive & negative)
+        Set<String> positiveInValues = new HashSet<>(); // contains values for positive in-filter
+        Set<String> negativeInValues = new HashSet<>(); // contains values for negative in-filter
+
         for (ApiFilter filter : filters) {
             ApiFilter normalizedFilter = filter;
             if (normalizedFilter.getOperation().equals(FilterOperation.notin)) {
                 normalizedFilter = filter.withOperation(FilterOperation.in);
             }
-            Filter disjunction = new OrFilter(buildSelectorFilters(
-                    dimension,
-                    getFilteredDimensionRows(dimension, Collections.singleton(normalizedFilter))
-            ));
 
-            // transform disjunction(OR) filter to in-filter
-            Filter inFilter = new InFilter(
-                    dimension,
-                    ((OrFilter) disjunction).getFields().stream()
-                            .map(subFilter -> (SelectorFilter) subFilter)
-                            .map(SelectorFilter::getValue)
-                            .distinct()
-                            .collect(Collectors.toList())
-            );
+            List<String> values = getFilteredDimensionRows(dimension, Collections.singleton(normalizedFilter)).stream()
+                    .map(row -> row.get(dimension.getKey()))
```

Comment:
DimensionRow has a function that does the same thing: `getKeyValue`. So you can just write:

```java
.map(DimensionRow::getKeyValue)
```